### PR TITLE
EIP-5521:  Referable NFT (rNFT)

### PR DESCRIPTION
**Pull request by OniReimu**
This standard proposes an extension to ERC-721 Non-Fungible Tokens (NFTs) to construct reference relationships among NFTs.

When opening a pull request to submit a new EIP, please use the suggested template: https://github.com/ethereum/EIPs/blob/master/eip-template.md

We have a GitHub bot that automatically merges some PRs. It will merge yours immediately if certain criteria are met:

 - The PR edits only existing draft PRs.
 - The build passes.
 - Your GitHub username or email address is listed in the 'author' header of all affected PRs, inside <triangular brackets>.
 - If matching on email address, the email address is the one publicly listed on your GitHub profile.

### CODE REVIEW

File: EIPS/eip-5521.md [hunk 1]
```diff
@@ -0,0 +1,187 @@
...
+*Incentive Compatibility*: This standard clarifies the referable relationship across different NFTs, helping to integrate multiple up-layer incentive models for both original NFT owners and new creators.
+
+*Easy Integration*: This standard makes it easier for the existing token standards or third-party protocols. For instance, the referable NFT can be applied to rentable scenarios (cf. [EIP-5006](./eip-5006.md) to build a hierarchical rental market, where multiple users can rent the same NFT during the same time or one user can rent multiple NFTs during the same duration). 
+
+## Backwards Compatibility
+This standard can be fully [EIP-721](./eip-721.md) compatible by adding an extension function set.
+
+## Test Cases
+Truffle and Openzeppelin are required to run the following in a test network.
+```
```
Comment (Pandapip1) [inline EIPS/eip-5521.md:52]:
```suggestion
```node
```

File: EIPS/eip-5521.md [hunk 2]
```diff
@@ -0,0 +1,187 @@
...
+
+rNFT.safeMint(3,[1,2])
+rNFT.referredOf(2)
+rNFT.referredOf(3)
+rNFT.referringOf(3)
+
+```
+
+## Reference Implementation
+```
```
Comment (Pandapip1) [inline EIPS/eip-5521.md:71]:
```suggestion
```solidity
```

File: EIPS/eip-5521.md [hunk 3]
```diff
@@ -0,0 +1,187 @@
...
+// SPDX-License-Identifier: MIT
```
Comment (Pandapip1) [inline EIPS/eip-5521.md:72]:
```suggestion
// SPDX-License-Identifier: CC0-1.0
```

File: EIPS/eip-5521.md [hunk 4]
```diff
@@ -0,0 +1,187 @@
...
+    function referringOf(uint256 tokenId) external view returns(uint256[] memory);
+
+    /// @notice Get the list of the rNFT that refers to `tokenId`
+    /// Throws if `tokenId` is not valid rNFT
+    /// @param tokenId The rNFT of the referred list
+    function referredOf(uint256 tokenId) external view returns(uint256[] memory);
+}
+```
+
+```
```
Comment (Pandapip1) [inline EIPS/eip-5521.md:98]:
```suggestion
```solidity
```

File: EIPS/eip-5521.md [hunk 5]
```diff
@@ -0,0 +1,187 @@
...
+// SPDX-License-Identifier: MIT
```
Comment (Pandapip1) [inline EIPS/eip-5521.md:99]:
```suggestion
// SPDX-License-Identifier: CC0-1.0
```

File: EIPS/eip-5521.md [hunk 6]
```diff
@@ -0,0 +1,187 @@
...
+
+    /// @dev See {IERC165-supportsInterface}.
+    function supportsInterface(bytes4 interfaceId) public view virtual override returns (bool) {
+        return interfaceId == type(IERC_rNFT).interfaceId || super.supportsInterface(interfaceId);
+    }
+}
+```
+
+## Security Considerations
+The *createdTimestamp* only covers the block-level timestamp (based on block headers), which does not support fine-grained comparison such as transaction-level.
```
Comment (Pandapip1) [inline EIPS/eip-5521.md:184]:
```suggestion
The `createdTimestamp` only covers the block-level timestamp (based on block headers), which does not support fine-grained comparison such as transaction-level.
```

File: EIPS/eip-5521.md [hunk 7]
```diff
@@ -0,0 +1,187 @@
...
+eip: 5521
+title: EIP-N Referable NFT
+description: EIP-N, Interface extension for 721, a referable version of NFT (rNFT)
+author: Saber Yu (@OniReimu), Qin Wang <qin.wang@data61.csiro.au>, Shange Fu <shange.fu@monash.edu>, Shiping Chen <shiping.chen@data61.csiso.au>, Sherry Xu <xiwei.xu@data61.csiro.au>, Jiangshan Yu <jiangshan.yu@monash.edu>
+discussions-to: https://ethereum-magicians.org/t/eip-x-erc-721-referable-nft/10310
+status: Draft
+type: Standards Track
+category: ERC
+created: 2022-08-10
+requires: 165, 721
```
Comment (Pandapip1) [inline EIPS/eip-5521.md:11]:
```suggestion
requires: 165, 721, 5006
```

**Review by Pandapip1 — CHANGES_REQUESTED** [review]

**Comment by OniReimu** [comment]
Implemented your suggestions and looking forward to the next review!

Thanks for your time and help @Pandapip1 :-)

**Review by Pandapip1 — APPROVED** [review]
Looks fine. Needs the codespell update

**Pull request merged**